File: src/main/java/algo/stack/QueueTwoStacks.java
package algo.stack;

import org.junit.Test;

import java.util.NoSuchElementException;
import java.util.Stack;

import static org.junit.Assert.assertTrue;

public class QueueTwoStacks {

    private Stack<Integer> inStack = new Stack<>();
    private Stack<Integer> outStack = new Stack<>();

    public void enqueue(int item) {
        inStack.push(item);
    }

    public int dequeue() {
        if (outStack.empty()) {

            // Move items from inStack to outStack, reversing order
            while (!inStack.empty()) {
                int newestInStackItem = inStack.peek();
                outStack.push(newestInStackItem);
                inStack.pop();
            }

            // If outStack is still empty, raise an error
            if (outStack.empty()) {
                throw new NoSuchElementException("Can't dequeue from empty queue!");
            }
        }
        return outStack.pop();
    }


    @Test
    public  void testDequeue() throws Exception {
        QueueTwoStacks queue = new QueueTwoStacks();
        queue.enqueue(5);
        queue.enqueue(4);
        queue.enqueue(3);
        queue.enqueue(2);

        assertTrue(5 == queue.dequeue());
        assertTrue(4 == queue.dequeue());
        assertTrue(3 == queue.dequeue());
        assertTrue(2 == queue.dequeue());

    }


}

File: src/main/python/queues/QueueTwoStacks.py
class QueueTwoStacks(object):
    def __init__(self):
        self.in_stack = []
        self.out_stack = []

    def enqueue(self, item):
        self.in_stack.append(item)

    def dequeue(self):
        if len(self.out_stack) == 0:

            # Move items from in_stack to out_stack, reversing order
            while len(self.in_stack) > 0:
                self.out_stack.append(self.in_stack.pop())

            # If out_stack is still empty, raise an error
            if len(self.out_stack) == 0:
                raise IndexError("Can't dequeue the empty queue")

        return self.out_stack.pop()

if __name__ == '__main__':
    queue = QueueTwoStacks();
    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(40)
    queue.enqueue(80)
    queue.enqueue(20)
    print(queue.in_stack)
    print(queue.dequeue())
    print(queue.dequeue())
    print(queue.dequeue())
